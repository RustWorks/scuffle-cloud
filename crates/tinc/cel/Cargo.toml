[package]
name = "tinc-cel"
version = "0.1.0"
edition = "2024"
description = "A CEL Type used by tinc-build and tinc for validating protobuf messages"
repository = "https://github.com/scufflecloud/scuffle"
authors = ["Scuffle <opensource@scuffle.cloud>"]
readme = "README.md"
documentation = "https://docs.rs/tinc-build"
license = "MIT OR Apache-2.0"
keywords = ["grpc", "protobuf", "tonic", "codegen"]

[dependencies]
num-traits = "0.2.19"
bytes = "1.5"
float-cmp = "0.10"
thiserror = "2"
chrono = "0.4"
regex = "1"
linkme = "0.3"
fmtools = "0.1"
uuid = "1"
url = "2"
email_address = "0.2.9"
scuffle-workspace-hack.workspace = true

[features]
runtime = []

