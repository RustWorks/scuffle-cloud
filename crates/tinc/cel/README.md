<!-- cargo-sync-rdme title [[ -->
# tinc-cel
<!-- cargo-sync-rdme ]] -->

> [!WARNING]  
> This crate is under active development and may not be stable.

<!-- cargo-sync-rdme badge [[ -->
![License: MIT OR Apache-2.0](https://img.shields.io/crates/l/tinc-cel.svg?style=flat-square)
[![docs.rs](https://img.shields.io/docsrs/tinc-cel.svg?logo=docs.rs&style=flat-square)](https://docs.rs/tinc-cel)
[![crates.io](https://img.shields.io/crates/v/tinc-cel.svg?logo=rust&style=flat-square)](https://crates.io/crates/tinc-cel)
[![GitHub Actions: ci](https://img.shields.io/github/actions/workflow/status/scufflecloud/scuffle/ci.yaml.svg?label=ci&logo=github&style=flat-square)](https://github.com/scufflecloud/scuffle/actions/workflows/ci.yaml)
[![Codecov](https://img.shields.io/codecov/c/github/scufflecloud/scuffle.svg?label=codecov&logo=codecov&style=flat-square)](https://codecov.io/gh/scufflecloud/scuffle)
<!-- cargo-sync-rdme ]] -->

---

<!-- cargo-sync-rdme rustdoc [[ -->
Currently this is a fully private api used by `tinc` and `tinc-build` to
compile and execute [CEL](https://cel.dev/) expressions.
Check out the [changelog](./CHANGELOG.md).

### Feature flags

* **`runtime`** —  Enables runtime enum type indexing
* **`docs`** —  Enables changelog and documentation of feature flags

### License

This project is licensed under the MIT or Apache-2.0 license.
You can choose between one of them if you use this work.

`SPDX-License-Identifier: MIT OR Apache-2.0`
<!-- cargo-sync-rdme ]] -->
