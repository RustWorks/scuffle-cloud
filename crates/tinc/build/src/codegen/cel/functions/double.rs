use syn::parse_quote;
use tinc_cel::CelValue;

use super::Function;
use crate::codegen::cel::compiler::{CompileError, CompiledExpr, CompilerCtx, ConstantCompiledExpr, RuntimeCompiledExpr};
use crate::codegen::cel::types::CelType;

#[derive(Debug, Clone, Default)]
pub(crate) struct Double;

impl Function for Double {
    fn name(&self) -> &'static str {
        "double"
    }

    fn syntax(&self) -> &'static str {
        "<this>.double()"
    }

    fn compile(&self, ctx: CompilerCtx) -> Result<CompiledExpr, CompileError> {
        let Some(this) = ctx.this else {
            return Err(CompileError::syntax("missing this", self));
        };

        if !ctx.args.is_empty() {
            return Err(CompileError::syntax("takes no arguments", self));
        }

        match this.into_cel()? {
            CompiledExpr::Constant(ConstantCompiledExpr { value }) => {
                Ok(CompiledExpr::constant(CelValue::cel_to_double(value)?))
            }
            CompiledExpr::Runtime(RuntimeCompiledExpr { expr, .. }) => Ok(CompiledExpr::runtime(
                CelType::CelValue,
                parse_quote!(::tinc::__private::cel::CelValue::cel_to_double(#expr)?),
            )),
        }
    }
}

#[cfg(test)]
#[cfg_attr(coverage_nightly, coverage(off))]
mod tests {
    use quote::quote;
    use syn::parse_quote;
    use tinc_cel::CelValue;

    use crate::codegen::cel::compiler::{CompiledExpr, Compiler, CompilerCtx};
    use crate::codegen::cel::functions::{Double, Function};
    use crate::codegen::cel::types::CelType;
    use crate::types::{ProtoType, ProtoTypeRegistry, ProtoValueType};

    #[test]
    fn test_bytes_syntax() {
        let registry = ProtoTypeRegistry::new();
        let compiler = Compiler::new(&registry);
        insta::assert_debug_snapshot!(Double.compile(CompilerCtx::new(compiler.child(), None, &[])), @r#"
        Err(
            InvalidSyntax {
                message: "missing this",
                syntax: "<this>.double()",
            },
        )
        "#);

        insta::assert_debug_snapshot!(Double.compile(CompilerCtx::new(compiler.child(), Some(CompiledExpr::constant(CelValue::String("13.2".into()))), &[])), @r"
        Ok(
            Constant(
                ConstantCompiledExpr {
                    value: Number(
                        F64(
                            13.2,
                        ),
                    ),
                },
            ),
        )
        ");

        insta::assert_debug_snapshot!(Double.compile(CompilerCtx::new(compiler.child(), Some(CompiledExpr::constant(CelValue::List(Default::default()))), &[
            cel_parser::parse("1 + 1").unwrap(), // not an ident
        ])), @r#"
        Err(
            InvalidSyntax {
                message: "takes no arguments",
                syntax: "<this>.double()",
            },
        )
        "#);
    }

    #[test]
    fn test_double_runtime() {
        let registry = ProtoTypeRegistry::new();
        let compiler = Compiler::new(&registry);

        let string_value =
            CompiledExpr::runtime(CelType::Proto(ProtoType::Value(ProtoValueType::String)), parse_quote!(input));

        let result = Double
            .compile(CompilerCtx::new(compiler.child(), Some(string_value), &[]))
            .unwrap();

        let small_fn = quote! {
            #[allow(dead_code)]
            fn double_conv(input: &std::string::String) -> Result<::tinc::__private::cel::CelValue<'_>, ::tinc::__private::cel::CelError<'_>> {
                Ok(#result)
            }
        };

        let compiled = postcompile::compile_str!(&small_fn.to_string());
        insta::assert_snapshot!(compiled);
    }
}
