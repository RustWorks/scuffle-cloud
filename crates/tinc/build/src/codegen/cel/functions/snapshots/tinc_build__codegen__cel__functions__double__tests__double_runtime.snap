---
source: crates/tinc/build/src/codegen/cel/functions/double.rs
expression: result
---
Runtime(
    RuntimeCompiledExpr {
        ty: CelValue,
        expr: ::tinc::__private::cel::CelValue::cel_to_double(
            ::tinc::__private::cel::CelValueConv::conv(input),
        )?,
    },
)
